torchao/experimental/ops/linear_8bit_act_xbit_weight/kernel_selector.h

```diff
@@ -0,0 +1,361 @@
+                                 has_clamp>}}}});
+      return;
+    }
+#endif // defined(__aarch64__) || defined(__ARM_NEON)
+  }
+}
+
+#if defined(TORCHAO_ENABLE_KLEIDI)
+template <typename kernel_struct, int m_step, int mr, int n_step, int nr,
+          int kr, int sr>
+UKernelConfig::linear_config_type get_linear_config_kleidi() {
+  namespace op = torchao::kernels::cpu::aarch64::kleidi::
+      kai_matmul_clamp_f32_qai8dxp_qsi4c32p;
+  assert(m_step == kernel_struct::get_ukernel().get_m_step());
+  assert(mr == kernel_struct::get_ukernel().get_mr());
+  assert(n_step == kernel_struct::get_ukernel().get_n_step());
+  assert(nr == kernel_struct::get_ukernel().get_nr());
+  assert(kr == kernel_struct::get_ukernel().get_kr());
+  assert(sr == kernel_struct::get_ukernel().get_sr());
+  return UKernelConfig::linear_config_type{
+      /*mr*/ m_step,
+      /*activation_data_size_fn*/ &op::activation_data_size<mr, kr, sr>,
+      /*prepare_activation_data_fn*/ &op::prepare_activation_data<mr, kr, sr>,
+      /*kernel*/ &kernel_struct::kernel};
+}
+
+template <int nr, int kr, int sr>
+UKernelConfig::weight_packing_config_type get_weight_packing_config_kleidi() {
+  namespace op = torchao::kernels::cpu::aarch64::kleidi::
+      kai_matmul_clamp_f32_qai8dxp_qsi4c32p;
+  return UKernelConfig::weight_packing_config_type(
+      {/*weight_data_size_fn*/ &op::weight_data_size<nr, kr, sr>,
+       /*prepare_weight_data_fn*/ &op::prepare_weight_data<nr, kr, sr>});
+}
+
+template <int weight_nbit, bool has_weight_zeros>
+void register_ukernel_config_kleidi(UKernelConfigRegistrationTable &table,
+                                    PackedWeightsFormat format,
+                                    cpuinfo_uarch uarch) {
+  if (!cpuinfo_initialize()) {
+    throw std::runtime_error("Failed to initialize cpuinfo!");
+  }
+  check_format<weight_nbit, has_weight_zeros, /*has_bias*/ true>(
+      format, torchao::ops::PackedWeightsType::kleidi_ai);
+  namespace op = torchao::kernels::cpu::aarch64::kleidi::
+      kai_matmul_clamp_f32_qai8dxp_qsi4c32p;
+
+  if (format.nr == 8 && format.kr == 16 && format.sr == 2) {
+    constexpr int nr = 8;
+    constexpr int kr = 16;
+    constexpr int sr = 2;
+#if defined(TORCHAO_ENABLE_ARM_I8MM)
+    if (cpuinfo_has_arm_i8mm()) {
+      constexpr int n_step = 8;
+      table.register_ukernel_config(
+          format, uarch,
+          UKernelConfig{
+              /*preferred_alignment*/ op::get_preferred_alignement(),
+              /*nr*/ n_step,
+              /*weight_packing_config*/
+              get_weight_packing_config_kleidi<nr, kr, sr>(),
+              /*linear_configs*/
+              {{get_linear_config_kleidi<
+                  op::matmul_clamp_f32_qai8dxp4x8_qsi4c32p8x8_4x8x32_neon_i8mm,
+                  /*m_step*/ 4, /*mr*/ 4, n_step, nr, kr, sr>()}}});
+      return;
+    }
+#endif // TORCHAO_ENABLE_ARM_I8MM
+
+    if (cpuinfo_has_arm_neon_dot()) {
+      constexpr int n_step = 8;
```

Comment:
unfortunate that we can't do `get_n_step()` or `get_n_step(nr)` here :\

Reply:
why can you not?